locator/tablet_sharder.hh

```diff
@@ -29,6 +29,74 @@ private:
             _tmap = &_tm.tablets().get_tablet_map(_table);
         }
     }
+
+    std::optional<unsigned> get_shard(const tablet_replica_set& replicas, host_id host) const {
+        for (auto&& r : replicas) {
+            if (r.host == host) {
+                return r.shard;
+            }
+        }
+        return std::nullopt;
+    };
+
+    dht::shard_replica_set shard_for_writes(tablet_id tid, host_id host) const {
+        auto* trinfo = _tmap->get_tablet_transition_info(tid);
+        auto& tinfo = _tmap->get_tablet_info(tid);
+        dht::shard_replica_set shards;
+
+        auto push_from = [&](const tablet_replica_set& replicas) {
+            for (auto&& r: replicas) {
+                if (r.host == host) {
+                    shards.push_back(r.shard);
+                }
+            }
+        };
+
+        // See "Shard assignment stability" from doc/dev/topology-over-raft.md for explanation of logic.
+
+        if (trinfo && trinfo->pending_replica && trinfo->pending_replica->host == host) {
+            if (trinfo->transition == tablet_transition_kind::intranode_migration) {
```

Comment:
@tgrabiec how about `tablet_transition_kind::migration`?
Don't we need to consider `write_replica_set_selector` in a similar way?

Reply:
We specifically don't want tablet_transition_kind::migration to behave in this way, that's why this logic applies only to intranode migration. For regular migration we need shard stability across all stages of migration, which is explained here: https://github.com/scylladb/scylladb/blob/master/docs/dev/topology-over-raft.md?plain=1#L424